store/postgres/src/notification_listener.rs

```diff
@@ -102,6 +107,34 @@ impl NotificationListener {
         Arc<AtomicBool>,
         Arc<Barrier>,
     ) {
+        /// Connect to the database at `postgres_url` and do a `LISTEN
+        /// {channel_name}`. If that fails, retry with exponential backoff
+        /// with a delay between 1s and 32s
+        fn connect_and_listen(logger: &Logger, postgres_url: &str, channel_name: &str) -> Client {
+            let mut attempt: usize = 0;
+            loop {
+                let res = Client::connect(postgres_url, NoTls).and_then(|mut conn| {
+                    conn.execute(format!("LISTEN {}", channel_name).as_str(), &[])?;
+                    Ok(conn)
+                });
+                match res {
+                    Err(e) => {
+                        let delay = Duration::from_secs(1 << attempt);
+                        error!(logger, "Failed to connect notification listener: {}", e;
+                                       "attempt" => attempt,
+                                       "retry_delay_s" => delay.as_secs());
+                        if attempt < 5 {
+                            attempt = attempt + 1;
+                        }
+                        thread::sleep(delay);
+                    }
+                    Ok(conn) => {
```

Comment:
The `return` is needed to get out of the outer `loop` - that also makes using `map_err` etc. awkward, specially since we just log the error and forget about it.